### PR TITLE
Feat: webhook reject unknown cr outputs

### LINKED ISSUE #6831: [Feature] Update Validating Webhook to reject Custom Definitions if the resources that it creates/manages does not exists on the cluster.
**Is your feature request related to a problem? Please describe.**
Currently the custom definitions(component definition, policy definition etc) are validated by the webhook if the workload type does not exist on the cluster. There is no validation on the outputs or output. This enables developers to create definition which have the ability create/manage resources(CRs) that do not exist on the cluster. This may lead to consumption of these definitions through applications and cause failures then. The validationg webhook should act as a fail fast mechanism during the definition creation phase itself.

**Describe the solution you'd like**
Update the validating webhooks to allow definitions which are compatible with the clusters. A definition containing a CR object in the outputs or output should be rejected.

### PR DESCRIPTION
**Pull request by roguepikachu**
### Description of your changes

Fixes #6831

Description of your changes

This PR introduces webhook validation for resource existence in CRD definitions, ensuring that resources referenced in CUE templates actually exist in the cluster before allowing the definition to be created or updated.

Key Changes

New Feature: Resource Existence Validation

- Added ValidateResourcesExist feature gate to enable webhook validation that checks if resource types referenced in ComponentDefinition, TraitDefinition, PolicyDefinition, and WorkflowStepDefinition CUE templates exist in the cluster
- Implemented comprehensive validation logic that parses CUE templates and validates output resources against the cluster's available APIs
- Enhanced error messages to clearly indicate which resource types are missing when validation fails

Webhook Handler Refactoring

- Refactored all definition validating handlers to use structured logging
with consistent field names
- Renamed webhook handlers for clarity (e.g., validating_handler.go →
component_definition_validating_handler.go)
- Improved error handling and logging throughout webhook validation process
- Added detailed logging at each validation step for better debugging

Enhanced Testing & Debugging

- Added comprehensive e2e tests (definition_output_validation_test.go) with covering various validation scenarios
- Created webhook debugging documentation (docs/WEBHOOK_DEBUGGING.md) with setup instructions
- Added debug webhook setup script (hack/debug-webhook-setup.sh) for local development
- Introduced VS Code launch configuration for webhook debugging

Infrastructure Improvements

- Added e2e-test-local make target for k3d cluster setup and webhook validation testing
- Enhanced addon definition detection using owner references for more accurate validation
- Updated Ingress API version to v1 for compatibility with modern Kubernetes clusters
- Added structured logging utility (pkg/logging/logger.go) for consistent log formatting

Helm Chart Updates

- Added webhook configuration for WorkflowStepDefinition validation
- Updated values.yaml with new feature gate configuration

Impact

This change improves the reliability of KubeVela by preventing the creationof definitions that reference non-existent resources, which would otherwise fail at runtime. The feature is disabled by default and can be enabled via the ValidateResourcesExist feature gate.

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

Testing

- All existing tests pass
- Added extensive e2e tests covering:
- Valid and invalid resource references in CUE templates
- Cross-namespace resource validation
- Addon vs user-defined definition differentiation
- Edge cases and error scenarios

### CODE REVIEW

File: charts/vela-core/README.md
Comment (roguepikachu) [inline charts/vela-core/README.md:1]: Adds a feature gate to validate the resource's existence mentioned in output/outputs for definitions before they are applied.

File: charts/vela-core/templates/admission-webhooks/validatingWebhookConfiguration.yaml
Comment (roguepikachu) [inline charts/vela-core/templates/admission-webhooks/validatingWebhookConfiguration.yaml:1]: include workflow definitions for the webhook validation

**Comment by roguepikachu** [comment]
### Webhook Validation for Unknown Resources

**Problem**

* Previously, users could create `ComponentDefinitions` / `TraitDefinitions`/ `PolicyDefinitions` / `WorkflowDefinitions`  referencing non-existent resource types (e.g., a CRD not yet installed).
* These would be accepted but then fail at runtime when the controller attempted to create the resource, causing confusion and runtime errors.

**Solution**

* Introduced webhook validation that inspects the CUE template outputs and checks against the cluster’s REST mapper to ensure the referenced resources actually exist.
* If the resource type cannot be resolved, the definition is rejected at admission time.

**Key Architectural Decisions**

* Feature-gated with a `ValidateResourcesExist` flag (disabled by default for backward compatibility).
* Validation occurs at admission time by parsing the CUE template and resolving outputs against the cluster’s API resources.
* Addon-provided definitions are checked for and skipped (via owner references, labels, or annotations) to prevent false positives during addon installation.
* Structured logging added throughout to aid in debugging and traceability.

**Edge Cases Handled**

* **Addon Definitions**: Skipped from validation since addons typically install CRDs and definitions together, which could otherwise trigger false failures (chicken-egg scenario ex installing FluxCD addon).
* **Conditional Outputs**: If a CUE template’s output or outputs block contains conditional rendering, the actual branch chosen is only known at runtime. Because of this, it’s difficult to fully validate such cases at webhook admission time. The validation logic makes a best effort, but runtime-only conditions may still escape strict pre-checking.

**Additional Enhancements**

* Refactored webhook handlers for consistent naming and structured logging.
* Added new e2e tests covering validation scenarios and edge cases.
* Added debugging documentation and local test setup scripts.
* Updated to use the `Ingress` v1 API for modern Kubernetes compatibility.

**Impact**

* Prevents runtime failures by catching invalid definitions at admission time.
* Provides safer, clearer feedback to users before their definitions are accepted.

File: hack/debug-webhook-setup.sh [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+    # Generate CA certificate
+    openssl req -x509 -new -nodes -key ${CERT_DIR}/ca.key -days 365 -out ${CERT_DIR}/ca.crt \
+        -subj "/CN=webhook-ca"
+
+    # Generate server private key
+    openssl genrsa -out ${CERT_DIR}/tls.key 2048
+
+    # Get host IP for Docker internal network
+    HOST_IP="192.168.5.2"
```
Comment (anoop2811) [inline hack/debug-webhook-setup.sh:82]: Should we us something like `docker network inspect bridge \
  | jq -r '.[0].IPAM.Config[0].Gateway'` for host ip?

File: makefiles/e2e.mk [hunk 2]
```diff
@@ -85,6 +87,32 @@ e2e-test:
 	ginkgo -v ./test/e2e-test
 	@$(OK) tests pass
 
+# Run e2e tests with k3d and webhook validation
+.PHONY: e2e-test-local
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply makefiles/e2e.mk:91]: Added the quotes

File: pkg/logging/logger.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+	if len(startTime) > 0 {
+		duration := time.Since(startTime[0])
+		logger = logger.WithValues(FieldDuration, duration.Milliseconds())
+	}
+	return logger
+}
+
+// WithError adds error context to the logger
+func (l Logger) WithError(err error) Logger {
+	return l.WithValues(FieldError, err.Error(), FieldSuccess, false)
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/logging/logger.go:147]: The method WithError is supposed to called with an error, and when err!=nil should be called upstream

File: makefiles/develop.mk [hunk 4]
```diff
@@ -30,3 +30,71 @@ core-run: fmt vet manifests
...
+	@echo "  make k3d-create               # Create k3d cluster"
+	@echo "  make k3d-delete               # Delete k3d cluster"
+	@echo "  make webhook-setup            # Setup webhook (certs + config)"
+	@echo "  make webhook-clean            # Clean up webhook setup"
+
+## k3d-create: Create a k3d cluster for debugging
+.PHONY: k3d-create
+k3d-create:
+	@echo "Creating k3d cluster: $(K3D_CLUSTER_NAME)"
+	@k3d cluster create $(K3D_CLUSTER_NAME) \
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply makefiles/develop.mk]: Added the quotes

File: pkg/webhook/core.oam.dev/v1beta1/policydefinition/policy_definition_validating_handler.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+			"hasSchematic", obj.Spec.Schematic != nil,
+			"version", obj.Spec.Version)
+
+		if obj.Spec.Schematic != nil && obj.Spec.Schematic.CUE != nil {
+			logger.WithStep("validate-cue").Info("Validating CUE template syntax and semantics for PolicyDefinition schematic")
+			if err := webhookutils.ValidateCueTemplate(obj.Spec.Schematic.CUE.Template); err != nil {
+				logger.WithStep("validate-cue").WithError(err).Error(err, "CUE template contains syntax errors or invalid constructs - template compilation failed")
+				return admission.Denied(fmt.Sprintf("%s (requestUID=%s)", err.Error(), req.UID))
+			}
+			if err := webhookutils.ValidateOutputResourcesExist(obj.Spec.Schematic.CUE.Template, h.Client.RESTMapper(), obj); err != nil {
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/webhook/core.oam.dev/v1beta1/policydefinition/policy_definition_validating_handler.go:86]: The DoS risk is minimal because the mapper.RESTMapping() call operates against locally cached Kubernetes API discovery data (not network calls), the webhook already has Kubernetes-enforced timeouts (10-30 seconds), and creating PolicyDefinitions requires admin-level RBAC permissions that aren't typically exposed to untrusted users.

File: pkg/webhook/core.oam.dev/v1beta1/policydefinition/policy_definition_validating_handler.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+		}
+		logger.WithStep("decode").Info("Successfully decoded PolicyDefinition from admission request",
+			"definitionName", obj.Name,
+			"namespace", obj.Namespace,
+			"hasSchematic", obj.Spec.Schematic != nil,
+			"version", obj.Spec.Version)
+
+		if obj.Spec.Schematic != nil && obj.Spec.Schematic.CUE != nil {
+			logger.WithStep("validate-cue").Info("Validating CUE template syntax and semantics for PolicyDefinition schematic")
+			if err := webhookutils.ValidateCueTemplate(obj.Spec.Schematic.CUE.Template); err != nil {
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/webhook/core.oam.dev/v1beta1/policydefinition/policy_definition_validating_handler.go:82]: same as the above

File: makefiles/e2e.mk [hunk 7]
```diff
@@ -85,6 +87,32 @@ e2e-test:
...
+e2e-test-local:
+	# Create k3d cluster if needed
+	@k3d cluster create kubevela-debug --servers 1 --agents 1 || true
+	# Build and load image
+	docker build -t vela-core:e2e-test -f Dockerfile . --build-arg=VERSION=e2e-test --build-arg=GITVERSION=test
+	k3d image import vela-core:e2e-test -c kubevela-debug
+	# Deploy with Helm
+	kubectl delete validatingwebhookconfiguration kubevela-vela-core-admission 2>/dev/null || true
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply makefiles/e2e.mk:99]: When k3d cluster is created in the previous line, ti automatically sets the context, besides this is for local development.

File: test/e2e-test/definition_output_validation_test.go [hunk 8]
```diff
@@ -0,0 +1,1188 @@
...
+	}
+}`,
+						},
+					},
+				},
+			}
+
+			err := k8sClient.Create(ctx, componentDef)
+			Expect(err).To(HaveOccurred())
+			Expect(err.Error()).To(ContainSubstring("resource type not found on cluster"))
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply test/e2e-test/definition_output_validation_test.go:121]: In the GitHub Actions workflow, the e2e-setup-core target is used, which ultimately calls e2e-setup-core-wo-auth (line 72), and this DOES enable the feature gate at line 34: --set featureGates.validateResourcesExist=true. So this is taken care off and in local the devs are expected to run `make e2e-test-local` which also enables this feature gate.

File: pkg/cue/definition/template_test.go [hunk 9]
```diff
@@ -781,7 +786,7 @@ parameter: {
...
-				"t3ingress": &unstructured.Unstructured{Object: map[string]interface{}{"apiVersion": "networking.k8s.io/v1beta1", "kind": "Ingress", "labels": map[string]interface{}{"config": "enemies-data"}, "metadata": map[string]interface{}{"name": "test"}, "spec": map[string]interface{}{"rules": []interface{}{map[string]interface{}{"host": "example.com", "http": map[string]interface{}{"paths": []interface{}{map[string]interface{}{"backend": map[string]interface{}{"serviceName": "test", "servicePort": int64(1080)}, "path": "ping"}}}}}}}},
+				"t3ingress": &unstructured.Unstructured{Object: map[string]interface{}{"apiVersion": "networking.k8s.io/v1", "kind": "Ingress", "labels": map[string]interface{}{"config": "enemies-data"}, "metadata": map[string]interface{}{"name": "test"}, "spec": map[string]interface{}{"rules": []interface{}{map[string]interface{}{"host": "example.com", "http": map[string]interface{}{"paths": []interface{}{map[string]interface{}{"path": "ping", "pathType": "Prefix", "backend": map[string]interface{}{"service": map[string]interface{}{"name": "test", "port": map[string]interface{}{"number": int64(1080)}}}}}}}}}}},
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/cue/definition/template_test.go:789]: added the slash for the path

File: pkg/features/controller_features.go [hunk 10]
```diff
@@ -119,6 +119,10 @@ const (
 
 	// EnableApplicationStatusMetrics enable the collection and export of application status metrics and structured logging
 	EnableApplicationStatusMetrics = "EnableApplicationStatusMetrics"
+
+	// ValidateResourcesExist enables webhook validation to check if resource types referenced in
+	// ComponentDefinition/TraitDefinition/WorkflowDefinition/PolicyDefinition CUE templates exist in the cluster
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply pkg/features/controller_features.go]: Updated the comment, mb

File: makefiles/develop.mk [hunk 11]
```diff
@@ -30,3 +30,71 @@ core-run: fmt vet manifests
...
+		--servers 1 \
+		--agents 1 \
+		--wait || (echo "k3d cluster already exists"; exit 0)
```
[the remark this answers is not in the document]
Comment (roguepikachu) [reply makefiles/develop.mk]: Fixed to check if the cluster already exist

File: hack/debug-webhook-setup.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (roguepikachu) [reply hack/debug-webhook-setup.sh:82]: The file hack/debug-webhook-setup.sh is explicitly a debugging tool, not for production. The hardcoded IP `192.168.5.2` is  the standard k3d host gateway IP that k3d uses to allow containers to reach the host machine.

This setup is meant for developers running webhooks locally while the Kubernetes cluster runs in k3d, a common development pattern. Added comment in the script as well.

**Review by anoop2811 — APPROVED** [review]
